colossalai/zero/low_level/bookkeeping/tensor_bucket.py

```diff
@@ -40,12 +47,33 @@ def get_bucket(self):
 
     def empty(self):
         self._bucket = []
-        self._size = 0
+        self._current_size = 0
+        self._write_back_pairs = {}
 
     def flatten(self):
         return _flatten_dense_tensors(self._bucket)
 
+    def unflatten(self, flat_tensor):
+        return _unflatten_dense_tensors(flat_tensor, self._bucket)
+
     def unflatten_and_copy(self, flat_tensor):
-        unflattened_tensor_list = _unflatten_dense_tensors(flat_tensor, self._bucket)
+        unflattened_tensor_list = self.unflatten(flat_tensor)
         for old, new in zip(self._bucket, unflattened_tensor_list):
             old.copy_(new)
+
+    def all_gather(self, group=None):
+        flat = self.flatten()
+        buffers = [torch.empty_like(flat) for _ in range(dist.get_world_size(group))]
+        dist.all_gather(buffers, flat, group=group)
+        unflat_buffers = [self.unflatten(buffer) for buffer in buffers]
+        # transpose the list of list
+        return list(map(list, zip(*unflat_buffers)))
+
+    def all_gather_(self, group=None):
```

Comment:
Perhaps you want to name these two methods more clearly?

Reply:
Do you have any suggestion?

Reply:
I think it might be just good to put the implementation of `all_gather` into `all_gather_` and name `all_gather_` as `all_gather`.

Reply:
done